File: lua/rational-numbers/rational-numbers.lua
local function gcd(a, b)
  return b == 0 and a or gcd(b, a % b)
end

local function obj_form(a)
  return { num = a[1], den = a[2] }
end

local function reduce(a)
  local r = obj_form(a)
  if r.num == 0 then
    return {0, 1}
  end
  if r.den < 0 then
    r.num = -1 * r.num
    r.den = -1 * r.den
  end
  local g = gcd(r.num, r.den)
  return {r.num / g, r.den / g}
end

local function add(a, b)
  local x = obj_form(a)
  local y = obj_form(b)
  return reduce({
    x.num * y.den + y.num * x.den,
    x.den * y.den
  })
end

local function negate(a)
  local x = obj_form(a)
  return { -1 * x.num, x.den }
end

local function subtract(a, b)
  return add(a, negate(b)) 
end

local function multiply(a, b)
  local x = obj_form(a)
  local y = obj_form(b)
  return reduce({ x.num * y.num, x.den * y.den })
end

local function invert(a)
  local x = obj_form(a)
  return { x.den, x.num }
end

local function divide(a, b)
  return multiply(a, invert(b))
end

local function abs(a)
  local x = obj_form(a)
  return {math.abs(x.num), math.abs(x.den) }
end

local function exp_rational(a, p)
  local x = obj_form(a)
  local b
  if p >= 0 then
    b = { x.num ^ p, x.den ^ p }
  else
    b = { x.den ^ (-1 * p), x.num ^ (-1 * p) }
  end
  return reduce(b)
end

local function exp_real(p, a)
  local x = obj_form(a)
  return p ^ (x.num / x.den)
end

return {
  add = add,
  subtract = subtract,
  multiply = multiply,
  divide = divide,
  abs = abs,
  exp_rational = exp_rational,
  exp_real = exp_real,
  reduce = reduce
}

File: c/rational-numbers/rational_numbers.c
#include "rational_numbers.h"
#include "my_math.h"
#include <stdlib.h>
#include <math.h>

static rational_t rationalize(int numerator, int denominator) {
    if (denominator < 0) {
        numerator *= -1;
        denominator *= -1;
    }
    int g = gcd(abs(numerator), denominator);
    return (rational_t){ numerator / g, denominator / g };
}

rational_t reduce(rational_t r) {
    return rationalize(r.numerator, r.denominator);
}

rational_t add(rational_t r1, rational_t r2) {
    int num = r1.numerator * r2.denominator + r2.numerator * r1.denominator;
    int denom = r1.denominator * r2.denominator;
    return rationalize(num, denom);
}

rational_t negate(rational_t r) {
    return rationalize(-r.numerator, r.denominator);
}

rational_t subtract(rational_t r1, rational_t r2) {
    return add(r1, negate(r2));
}

rational_t multiply(rational_t r1, rational_t r2) {
    int num = r1.numerator * r2.numerator;
    int denom = r1.denominator * r2.denominator;
    return rationalize(num, denom);
}

rational_t invert(rational_t r) {
    return rationalize(r.denominator, r.numerator);
}

rational_t divide(rational_t r1, rational_t r2) {
    return multiply(r1, invert(r2));
}

rational_t absolute(rational_t r) {
    return rationalize(abs(r.numerator), abs(r.denominator));
}

rational_t exp_rational(rational_t r, int16_t n) {
    if (n < 0)
        return exp_rational(invert(r), -n);

    int num = int_pow(r.numerator, n);
    int denom = int_pow(r.denominator, n);
    return rationalize(num, denom);
}

float exp_real(uint16_t x, rational_t r) {
    return nthroot(powf(x, r.numerator),
                   abs(r.denominator));
}
